### ISSUE TITLE
Support cache-dependency-paths outside the current directory

### ISSUE THREAD

**Issue by PeterJCLaw** [actions/setup-python#361]
**Description:**

I'm building a composite Action which uses `actions/setup-python` internally. When specifying the `requirements.txt` from the action setting up Python fails, though I thiiink I'm passing the right values.

Here's what I've got:
``` yaml
    - name: Set up Python
      uses: actions/setup-python@v3
      with:
        python-version: '3.10'
        cache: pip
        cache-dependency-path: ${{ github.action_path }}/requirements.txt
```

Here's the [error I get](https://github.com/srobo/sr2022-comp/runs/5711971554?check_suite_focus=true):
> No file in /home/runner/work/sr2022-comp/sr2022-comp matched to [/home/runner/work/_actions/PeterJCLaw/srcomp-validate-action/bd809ff12b77607cc2f0ebbd035d6f0a5c8d32fc/requirements.txt], make sure you have checked out the target repository

`sr2022-comp` is the repo I'm using to test the action, `srcomp-validate-action` is the action.

My specific use-case only has a couple of dependencies so #351 would likely work as a fix, though I can imagine that a more complex scenario would want this to work.

**Platform:** Ubuntu

**Comment by PeterJCLaw** [comment]
I'm currently working around this by coping the requirements file into a temporary directory within the working directory and then referencing that, instead of the original. This doesn't feel like a great solution though as it ends up polluting the target repo's working copy.

**Comment by dmitry-shibanov** [comment]
Hello @PeterJCLaw. Thank you for your report. Could you please provide a public repository to reproduce the issue ?

**Comment by dmitry-shibanov** [comment]
Hello @PeterJCLaw. Actually, I think the root cause of the issue in the [toolkit/glob](https://github.com/actions/toolkit/tree/main/packages/glob) package. The [hashFiles](https://github.com/actions/toolkit/blob/main/packages/glob/src/internal-hash-files.ts#L16-L19) function tries to compare if the passed directory matches with [GITHUB_WORKSPACE](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables). The composite action's directory and checkout project's directory don't match, that is why if does hash the file and the actions setup-python throws an error.

**Comment by PeterJCLaw** [comment]
Thanks for your replies.
For completeness the repos in question are:
- https://github.com/PeterJCLaw/srcomp-validate-action/
- https://github.com/srobo/sr2022-comp/ (specifically https://github.com/srobo/sr2022-comp/pull/5)

I've raised https://github.com/actions/toolkit/issues/1035 for the underlying issue.

**Comment by dmitry-shibanov** [comment]
Hello @PeterJCLaw. For now I'm going to close the pull request because the root cause of the issue comes from the toolkit repository. If you have any concerns feel free to ping us.

**Comment by PeterJCLaw** [comment]
@dmitry-shibanov would you be open to reopening this issue? It looks like `hashFiles` has a new parameter now which allows callers to specify the workspace directory (https://github.com/actions/toolkit/pull/1318), which I think would help here?

**Comment by dmitry-shibanov** [comment]
Hello @PeterJCLaw. Could you please create a separate issue that we could investigate it one more time.

**Comment by zyv** [comment]
@dmitry-shibanov we also got hit by this one, any news whether this can be addressed now, that there is a new parameter in toolkit available?

**Comment by jaimehrubiks** [comment]
I'm also hit by this issue, I guess this line is quite common:

    ${{ github.action_path }}

**Comment by zyv** [comment]
@dmitry-shibanov any progress on that?

**Comment by aparnajyothi-y** [comment]
Hello Everyone, Thank you for reporting this issue. The changes related to the new hashFiles parameter, which allows callers to specify the workspace directory, are part of [PR #1318](https://github.com/actions/toolkit/pull/1318) in the Toolkit and have been incorporated into setup-python as part of [PR #772](https://github.com/actions/setup-python/pull/772). These changes are included in setup-python v5.

The current limitation when using the setup-python action with composite actions arises from the fact that it supports relative paths, while absolute paths are not yet allowed. This could be considered a feature request rather than a bug.

To work around this limitation when using absolute paths, we create a symbolic link to requirements.txt from the composite action's directory. This allows the setup-python action to access it, even though the path would otherwise be relative to the repository's root.

```
# Step 2: Create symbolic link for requirements.txt
- name: Create symbolic link for requirements.txt
  run: ln -s ${{ github.action_path }}/requirements.txt
  shell: bash
```

Please feel free to reach out if you need any further clarifications or have additional concerns.

**Comment by dekimsey** [comment]
Hi @aparnajyothi-y, I'd like to ask for clarification in your example:
> To work around this limitation when using absolute paths, we create a symbolic link to requirements.txt from the composite action's directory. 

Would this not overwrite or otherwise modify the workspace just to execute the setup-python action?

It brings to mind all sorts of edge cases now about testing for existing requirements.txt, having a third step to clean it up, etc.
When I author actions (including composite) I tend to try to keep side-effect changes on disk to a temp directory or in my own scratch space (in the case of composites). I cannot guarantee what state the workspace is in before or after my action executes.

**Comment by zhimsel** [comment]
> To work around this limitation when using absolute paths, we create a symbolic link to requirements.txt from the composite action's directory.

> Would this not overwrite or otherwise modify the workspace just to execute the setup-python action?

You can _partially_ hedge this concern by naming your symlink something that wouldn't affect the "parent" repo at all. For example:

```diff
    - name: Create symlink for requirements.txt
-     run: ln -s ${{ github.action_path }}/requirements.txt
+     run: ln -s ${{ github.action_path }}/requirements.txt foobarbazboom.txt
      shell: bash

    - name: Set up Python
      uses: actions/setup-python
      with:
        python-version: '3.11'
        cache: 'pip'
+       cache-dependency-path: foobarbazboom.txt
```

And if desired, clean it up afterwards with a simple `rm`.

**Comment by JasonYang545** [comment]
What was the reasoning behind the design to not allow for using the `requirements.txt` file within the action's repository? Did this come from a GitHub Actions file context scoping issue or som other factor?

**Comment by v-aparnajyothi-y** [comment]
Hello Everyone,  we’ve raised https://github.com/actions/setup-python/pull/1128 to address the above request. Please validate it against the requirement that caching should work when the project is not located at the root of the repository, and confirm if we can proceed further.
Feel free to reach out if you have any questions or need additional clarification.

**Comment by v-aparnajyothi-y** [comment]
Hello Everyone, The issue regarding cache not working when the project is not located at the root of the repository has been addressed and resolved through [PR #1128](https://github.com/actions/setup-python/pull/1128), which has now been merged into main.

With this update, caching behavior should now work when the project is not in the root of the repository hence closing this issue. Please feel free reach us in case of any concerns / clarifications needed.

**Comment by PeterJCLaw** [comment]
@aparnajyothi-y apologies for not replying sooner, I think I missed the previous request for validation. Unfortunately I think there are some issues with #1128 which make it unsuitable as a fix for this -- mainly that is quite likely to result in accidental pollution of the end-user's working checkout, but also that it doesn't solve this issue for cases not involving a composite action. (Have posted there with more details, but happy to open a separate issue if that's preferred).

Edit: I would encourage that any fix which includes placement of the file within the workspace make use of https://nodejs.org/api/fs.html#fs_fs_mkdtemp_prefix_options_callback, as that will enable creation of a non-conflicting directory to place the file in.

**Comment by v-aparnajyothi-y** [comment]
Hello @PeterJCLaw, Using fs.mkdtemp() to copy a file into /tmp doesn’t help if a matching file exists in the workspace root, because tools like setup-python resolve cache-dependency-path globs from GITHUB_WORKSPACE. So, the file in /tmp is ignored unless explicitly referenced in the workflow YAML.

To address this, we default overwrite to false, preventing unintended overwrites of user files. If a file with the same name already exists in the workspace, we skip copying and display a clear warning. This avoids accidental pollution of the user’s repository while still allowing advanced users to opt in via overwrite: true when needed.

This approach ensures safety by default and flexibility when intentional overrides are required.
We’ve captured this logic in this [PR #1149](https://github.com/actions/setup-python/pull/1149). Please validate from your end and let us know in cases of any concerns/clarifications are required

**Comment by PeterJCLaw** [comment]
> Hello [@PeterJCLaw](https://github.com/PeterJCLaw), Using fs.mkdtemp() to copy a file into /tmp doesn’t help if a matching file exists in the workspace root, because tools like setup-python resolve cache-dependency-path globs from GITHUB_WORKSPACE. So, the file in /tmp is ignored unless explicitly referenced in the workflow YAML.

@aparnajyothi-y thanks for picking this up. Apologies if there was a lack of clarity on my part -- I was expecting that the use of `fs.mkdtemp()` would be to create a directory _within the workspace_, by passing in a suitable `prefix`.

This would ensure that the action always work without additional effort from callers. This is important since the immediate caller _is (probably) not the one in control of the workspace_ -- they're merely providing an action for _someone else_ to use. I don't think it's a great idea to force the author of the intermediate action (i.e: caller of `setup-python`) to have to decide whether or not to overwrite something in _their caller's_ workspace -- since they don't know what's there.

This approach does accept a small risk that the calling code might accidentally conflict with the directory created, however IMO this is vanishingly small and a worthwhile trade-off as a workaround until https://github.com/actions/toolkit/issues/1035 is addressed.

In case it's useful to see how this would work, while not implemented in JavaScript, this is the same as the workaround logic I've used in the action I authored which triggered this report. Here's the lines in question: https://github.com/PeterJCLaw/srcomp-validate-action/blob/15580edf2a3caf3c68d11ca159bb4d56a807cf49/action.yml#L12-L14

**Comment by v-aparnajyothi-y** [comment]
Hello @PeterJCLaw, Thanks for validating and sharing the details. As confirmed, dependency files outside $GITHUB_WORKSPACE (e.g., those written to /tmp using fs.mkdtemp()) are not picked up by setup-python's cache-dependency-path logic due to how glob resolution works.
For now, we'll proceed with the current PR that copies the dependency file into the workspace by default (with overwrite: false) to ensure cache correctness and prevent unintended file pollution.
We agree that using temp directories would be ideal once upstream support (e.g., toolkit#1035) is in place. We'll revisit and enhance the implementation when that becomes available. Let us know if you have any further suggestions!

**Comment by PeterJCLaw** [comment]
@aparnajyothi-y thanks for the reply, however I think there is a misunderstanding here and I wanted to make a few things really clear:

* I do not believe that the current fix (with or without [PR #1149](https://github.com/actions/setup-python/pull/1149)) addresses the issue that I originally raised -- I would need to continue working around this issue (as I already have) regardless of whether these changes are released.

* I believe that the fix in [PR #1128](https://github.com/actions/setup-python/pull/1128) as it stands is a net negative change due to the issue I raised around accidental overwriting, so I would discourage releasing `actions/setup-python` in the current state

* The changes in [PR #1149](https://github.com/actions/setup-python/pull/1149) are neither necessary nor sufficient to address the core issue here, yet does introduce additional complexity, so I would discourage merging that PR or releasing `actions/setup-python` with that merged

* My core concern with the approach the fixes take is that they are unsafe by default and fail quietly. In contrast the reported behaviour is at least a loud error, enabling the user to (fairly) easily diagnose and resolve the issue. With these changes, the failure mode is quieter and push additional burden onto users to work around something that this action can address directly.

* I am aware that a directory (temporary or otherwise) outside the workspace would not be picked up. Using such a directory is _not_ what I am proposing.

What I am proposing is a solution like this:
1. create a temporary directory (guaranteed to be uniquely named) _within_ the workspace -- my reading of the docs suggests that `fs.mkdtemp` supports this by passing the `prefix` argument
1. copy or link the necessary files into that directory

Is there a reason that this proposed solution would not work? What am I missing here?

**Comment by v-aparnajyothi-y** [comment]
Hi everyone, Following up on this discussion and the recent implementation in [PR #1128](https://github.com/actions/setup-python/pull/1128):

After further review and discussion, we’ve decided to revert [PR #1128](https://github.com/actions/setup-python/pull/1128). In that implementation, a file was copied from the composite action into the main workspace at a repository level. This was initially intended to support composite actions in setup-python and provided valuable insight into how such scenarios could be addressed.

Moving forward, we’ll be implementing this capability at the toolkit level, as tracked in [actions/toolkit#1035](https://github.com/actions/toolkit/issues/1035)
. This will allow us to provide consistent support for composite actions in a secure and scalable way, while also unblocking hashFiles path handling at its root.
